### DISCUSSION TITLE
How to apply middleware before any route

### DISCUSSION THREAD

**Discussion by engineerhead** [tokio-rs/axum#2271] [Q&A] [answered]
Hey!
I may have missed it in documentation but couldn't find a way to something straight  which can be done in express.js like the following code. 
`
app.use((req,res,next)=>{
  if(!req.cookies.currency){
    res.cookie("currency", "USD")
  }
  next()
})
`

**Comment by davidpdrsn** [comment, marked as answer]
https://docs.rs/axum/0.6.20/axum/middleware/index.html#rewriting-request-uri-in-middleware

**Reply by FabijanC** [reply]
Any way to also apply some logic to the response in case the request was mapped? With axum::middleware it's doable because calling `next.run()` separates the pre- and post-endpoint handler logic, but it's called after routing is resolved.

**Reply by jplatte** [reply]
The example David linked to uses `axum::middleware` so I don't understand: What have you tried?

**Reply by FabijanC** [reply]
Looking at the documentation for the latest version, it no longer uses `axum::middleware` and there is no way to separate between the Request and the Response phase since it doesn't have the `next` argument.

**Reply by jplatte** [reply]
[`axum::middleware::from_fn`](https://docs.rs/axum/latest/axum/middleware/fn.from_fn.html) still exists though and you can still use it. It's just easier for that specific example to use a `MapRequestLayer`. If you want to do something more advanced, you can keep using `from_fn`.
